pants

```diff
@@ -53,8 +55,17 @@ source ${HERE}/build-support/pants_venv
 # + bootstrap_native_code: Builds target-specific native engine binaries.
 source ${HERE}/build-support/bin/native/bootstrap_code.sh
 
-# Default to using Python 2.7 if not otherwise specified.
-export PY="${PY:-python2.7}"
+# Default to using Python 3 if not otherwise specified.
+export PY="${PY:-python3}"
+
+# Set interpreter constraints to exactly match the interpreter used for the venv, if not already set.
+# Note that $PY only impacts which virtualenv we use for the parent Pants process; we must also set
+# PANTS_PYTHON_SETUP_INTERPRETER_CONSTRAINTS to constrain spawned subprocesses such as tests. Without
+# any interpreter constraints, when PY is set to Python 3 we get the _Py_Dealloc exception (#6985) as
+# Pants will try to use Python 2 for subprocesses. Without setting minor version constraints, when using Python 3
+# we could end up using a different Python minor version for subprocesses than we do for Pants.
+py_major_minor=$(${PY} -c 'import sys; print(".".join(map(str, sys.version_info[0:2])))')
+export PANTS_PYTHON_SETUP_INTERPRETER_CONSTRAINTS="${PANTS_PYTHON_SETUP_INTERPRETER_CONSTRAINTS:-['CPython==${py_major_minor}.*']}"
```

Comment:
Note: here we allow any patch version, but in `ci.sh` we determined we must specify the patch version. 

This was discovered while working on https://github.com/pantsbuild/pants/pull/7235. There, we have system Python 2.7 already, but must install a new version of Python 2.7 to get different unicode settings (UCS2 vs UCS4). We were running into issues with the interpreter changing between what was used to run Pants and what we used for subprocesses.

Should we default to constraining to the exact patch version too? Because this constraint is determined dynamically at runtime, I would advocate yes. The dynamic determination means that my computer could be using 3.7.1 and another's could be using 3.7.2, for example, without any issues. All it ensures is that that Pants run uses the exact same interpreter for subprocesses that it used for $PY / the virtualenv.

Reply:
I doesn't feel like requiring a precise version in the `pants` script is strictly nec... since the tests themselves don't require a precise version, and since the tests should run under some stableish version, even if pants is running under something else.

But it doesn't seem like it hurts to force them to match, so if it's caused you issues in that regard, it seems fine to align them. It's also possible that requiring a precise version only in CI (via the ci.sh/travis.yml) would be sufficient.

Reply:
I ended up making the change to require the exact patch version because while it's most often not important which interpreter it uses for tests—so long as they're the same minor version—there is a distinct case this would matter: interpreters with different UCS settings (i.e. how they store unicode).

For example, assume we have Python 2.7.13 built with UCS4 and have 2.7.15 built with UCS2. There is a meaningful difference between 2.7.13 vs. 2.7.15 here, due to the unicode setting. So, here, it does matter that we always consistently use 2.7.13 completely or 2.7.15 completely, and do not mix the two. This is what is going on in our `ci.sh` for #7235.

I can't think of a situation where it would be wrong to not have this guarantee that `$PY` interpreter is the same interpreter for subprocesses. Especially because we have the safety valve of letting the user pre-define `PANTS_PYTHON_SETUP_INTERPRETER_CONSTRAINTS`, I think this is the more correct implementation.